packages/compiler/src/parser/state/tag.ts

```diff
@@ -0,0 +1,219 @@
+		value = readAttributeValue(parser);
+		end = parser.index;
+	} else if (parser.matchRegex(regex_starts_with_quote_characters)) {
+		parser.error(PARSER_ERRORS.unexpectedToken('='), parser.index);
+	}
+	checkUnique(name);
+	return {
+		start,
+		end,
+		type: 'Attribute',
+		name,
+		value
+	};
+}
+
+function readAttributeValue(parser: Parser) {
+	const quoteMark = parser.eat("'") ? "'" : parser.eat('"') ? '"' : null;
+	if (quoteMark && parser.eat(quoteMark)) {
+		return [
+			{
+				start: parser.index - 1,
+				end: parser.index - 1,
+				type: 'Text',
+				raw: '',
+				data: ''
+			}
+		];
+	}
+	let value: TemplateNode[] = [];
+	try {
+		value = readSequence(
+			parser,
+			() => {
+				// handle common case of quote marks existing outside of regex for performance reasons
+				if (quoteMark) return parser.match(quoteMark);
+				return !!parser.matchRegex(regexStartsWithInvalidAttrValue);
+			},
+			'in attribute value'
+		);
+	} catch (e) {
+		const error = e as CompileError;
+		if (error.code === 'parse-error') {
+			// if the attribute value didn't close + self-closing tag
+			// eg: `<Component test={{a:1} />`
+			// acorn may throw a `Unterminated regular expression` because of `/>`
+			if (parser.template.slice(error.pos! - 1, error.pos! + 1) === '/>') {
+				parser.index = error.pos!;
+				parser.error(PARSER_ERRORS.unclosedAttributeValue(quoteMark || CUSTOM_TAG_END));
+			}
+		}
+		throw error;
+	}
+	if (value.length === 0 && !quoteMark) {
+		parser.error(PARSER_ERRORS.missingAttributeValue);
+	}
+	if (quoteMark) parser.index += 1;
+	return value;
+}
+
+function readSequence(parser: Parser, done: () => boolean, location: string): TemplateNode[] {
+	let currentChunk: Text = {
+		start: parser.index,
+		end: null,
+		type: 'Text',
+		raw: '',
+		data: '',
+	};
+
+	const chunks: TemplateNode[] = [];
+
+	function flush(end: number) {
+		if (currentChunk.raw) {
+			currentChunk.data = decode_character_references(currentChunk.raw, true);
+			currentChunk.end = end;
+			chunks.push(currentChunk);
+		}
+	}
+	while (parser.index < parser.template.length) {
+		const index = parser.index;
+		if (done()) {
+			flush(parser.index);
+			return chunks;
+		} else if (parser.eat(CUSTOM_TAG_START)) {
+			if (parser.match('#')) {
+				const index = parser.index - 1;
+				parser.eat('#');
+				const name = parser.readUntil(/[^a-z]/);
+				parser.error(PARSER_ERRORS.invalidLogicBlockPlacement(location, name), index);
+			}
+			flush(parser.index - 1);
+			parser.allowWhitespace();
+			const expression = read_expression(parser);
+			parser.allowWhitespace();
+			parser.eat(CUSTOM_TAG_END, true);
+			chunks.push({
+				start: index,
+				end: parser.index,
+				type: 'MustacheTag',
+				expression
+			});
+			currentChunk = {
+				start: parser.index,
+				end: null,
+				type: 'Text',
+				raw: '',
+				data: ''
+			};
+		} else {
+			currentChunk.raw += parser.template[parser.index++];
+		}
+	}
+	parser.error(PARSER_ERRORS.unexpectedEof);
+}
```

Comment:
is this code copied from somewhere?

Reply:
don't worry when he rebase with main prettier will complain